### DISCUSSION TITLE
where does RollingFileAppender implement MakeWriterExt?

### DISCUSSION THREAD

**Discussion by keepsimple1** [tokio-rs/tracing#1844] [Q&A] [answered]
https://github.com/tokio-rs/tracing/blob/6f23c128fced6409008838a3223d76d7332d79e9/examples/examples/appender-multifile.rs#L17

From the above example code, `RollingFileAppender` implements `MakeWriterExt` trait and can use `and`. Does that mean any type that implement `std::io::Write` also implements `MakeWriterExt` when using `tracing-subscriber`? 

I tried a different 3rd-party [`RollingFileAppender`](https://docs.rs/rolling-file/0.1.0/rolling_file/struct.RollingFileAppender.html) that implements `std::io::Write` but seems not have `MakeWriterExt` implemented. I'm confused and wondering where `tracing_appender::rolling::RollingFileAppender` implements `MakeWriterExt` trait.

**Comment by davidbarsky** [comment, marked as answer]
> From the above example code, `RollingFileAppender` implements `MakeWriterExt` trait and can use `and`. Does that mean any type that implement `std::io::Write` also implements `MakeWriterExt` when using `tracing-subscriber`?

Not quite. `MakeWriter` is implemented on anything that implements `std::io::Write` and is wrapped by [a `std::sync::Arc` or `std::sync::Mutex`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/trait.MakeWriter.html#foreign-impls). The [documentation on `MakeWriter` notes](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/trait.MakeWriter.html):

> This trait is already implemented for function pointers and immutably-borrowing closures that return an instance of `io::Write`, such as `io::stdout` and `io::stderr`. Additionally, it is implemented for `std::sync::Mutex` when the type inside the mutex implements `io::Write`.

Transitively, `MakeWriterExt` is implemented [on types implementing `MakeWriter`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/writer/trait.MakeWriterExt.html#implementors). However, the use `.and` doesn't come from a blanket implementation; it comes from [this `MakeWriter` implementation](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/trait.MakeWriter.html#impl-MakeWriter%3C%27a%3E-5).
